src/Symfony/Component/Serializer/Debug/SerializerActionFactory.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Debug;
+
+use Symfony\Component\Serializer\Debug\Normalizer\Denormalization;
+use Symfony\Component\Serializer\Debug\Normalizer\Normalization;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+final class SerializerActionFactory implements SerializerActionFactoryInterface
+{
+    public function createDenormalization(DenormalizerInterface $denormalizer, $data, $result, string $type, string $format, array $context = []): Denormalization
+    {
+        return new Denormalization(
+            $denormalizer,
+            $this->sanitize($data),
+            $this->sanitize($result),
+            $type,
+            $format,
+            $context
+        );
+    }
+
+    public function createNormalization(NormalizerInterface $normalizer, $object, $result, string $format, array $context = []): Normalization
+    {
+        return new Normalization(
```

Comment:
Same as raised by @jderusse in https://github.com/symfony/symfony/pull/39325#discussion_r536840642, I'm not convinced about the need of an object to store these information for this specific use-case (yet I usually agree about not using arrays for structured data).

Also related to my comment about the factory interface, this factory is almost pure internal details of the traceable normalizers and serializer, to share some methods, so are these objects. The factory could be internal and instantiated from the traceable objects, or refactored as a trait.

if we'd like to keep these objects, I think at least:
- the factory interface should be removed
- the factory class and these objects must be `@internal`
- the factory can be instantiated from the traceable instances directly

WDYT?